src/software/geom/util.cpp

```diff
@@ -1251,6 +1116,172 @@ std::optional<Segment> mergeFullyOverlappingSegments(Segment segment1, Segment s
+    if (segments.size() == 0)
+    {
+        return std::nullopt;
+    }
+    else if (segments.size() == 1)
+    {
+        // If there is only 1 segments, it is unique and should be returned
+        return std::make_optional(segments);
+    }
+
+    // Check that all segments are collinear
+    for (unsigned int i = 0; i < segments.size() - 1; i++)
+    {
+        if (!collinear(segments[i], segments[i + 1]))
+        {
+            return std::nullopt;
+        }
+    }
+    std::vector<Segment> unique_segments;
+
+    unsigned int j = 0;
+    // Loop through all segments and combine segments
+    // to reduce the vector to the smallest number of independent (not overlapping)
+    // segments
+    while (segments.size() > 0)
+    {
+        std::optional<Segment> temp_segment = segments[0];
+        unique_segments.push_back(temp_segment.value());
+        segments.erase(segments.begin());
+
+        for (unsigned int i = 0; i < segments.size(); i++)
+        {
+            temp_segment =
+                mergeOverlappingParallelSegments(unique_segments[j], segments[i]);
+
+            if (temp_segment.has_value())
+            {
+                unique_segments[j] = temp_segment.value();
+                // Remove segments[i] from the list as it is not unique
+                segments.erase(segments.begin() + i);
+                i--;
+            }
+        }
+        j++;
+    }
+
+    return unique_segments;
+}
+
+Shot calcMostOpenDirection(Point origin, Segment segment, std::vector<Circle> obstacles)
+{
+    std::vector<Segment> obstacle_segment_projections;
+    // If there are no obstacles, return the center of the Segment and the shot angle
+    if (obstacles.size() == 0)
+    {
+        return Shot(Point::getAverageOfPoints({segment.getSegStart(), segment.getEnd()}),
+                    ((segment.getSegStart() - origin)
+                         .orientation()
+                         .minDiff((segment.getEnd() - origin).orientation()))
+                        .abs());
+    }
+
+    // Loop through all obstacles to create their 'blocking' Segment
+    for (Circle circle : obstacles)
+    {
+        // If the reference is inside an obstacle there is no open direction
+        if (contains(circle, origin))
+        {
+            return Shot(Point(segment.getSegStart() +
+                              (segment.getEnd() - segment.getSegStart()) / 2),
+                        Angle::fromDegrees(0));
+        }
+
+        // Get the tangent rays from the reference point to the obstacle
+        auto [ray1, ray2] = getCircleTangentRaysToReference(origin, circle);
+
+        // Project the tangent Rays to obtain a 'blocked' segment on the reference Segment
+        std::optional<Segment> intersect_segment =
+            getIntersectingSegment(ray1, ray2, segment);
+
+        if (intersect_segment.has_value())
+        {
+            obstacle_segment_projections.push_back(intersect_segment.value());
+        }
```

Comment:
What about the case where one ray intersects the segment, but the other ray is off to the side? In that case wouldn't there still be part of the segment blocked?

ie.
![Screenshot from 2019-12-23 10-30-50](https://user-images.githubusercontent.com/9075711/71374386-56700180-256f-11ea-90bc-4a0f80f5b981.png)
(red is the part of the segment I'd expect to be blocked)

Reply:
Yup, good drawing.

This case is handled by getIntersectingSegment()

If this happens, getIntersectingSegment() returns the Segment from the edge of the reference Segment to the single Point of intersection